File: api/bigrag/models/query.py
from __future__ import annotations

from pydantic import BaseModel, Field


class QueryRequest(BaseModel):
    query: str
    top_k: int | None = Field(default=None, ge=1, le=1000)
    filters: dict | None = None
    min_score: float | None = None
    search_mode: str | None = Field(default=None, pattern=r"^(semantic|keyword|hybrid)$")
    rerank: bool | None = None


class VectorEntry(BaseModel):
    id: str
    embedding: list[float]
    text: str = ""
    metadata: dict = {}


class VectorUpsertRequest(BaseModel):
    vectors: list[VectorEntry]


class VectorDeleteRequest(BaseModel):
    ids: list[str]


class VectorUpsertResponse(BaseModel):
    status: str = "ok"
    upserted: int


class VectorDeleteResponse(BaseModel):
    status: str = "ok"
    deleted: int


class QueryResult(BaseModel):
    id: str
    text: str
    score: float
    document_id: str | None = None
    document_filename: str | None = None
    chunk_index: int | None = None
    page_no: int | None = None
    char_start: int | None = None
    char_end: int | None = None
    metadata: dict = {}


class QueryTimings(BaseModel):
    embed_ms: float = 0.0
    search_ms: float = 0.0
    rerank_ms: float = 0.0
    cache_ms: float = 0.0
    total_ms: float = 0.0
    cache_hit: bool = False


class QueryResponse(BaseModel):
    results: list[QueryResult]
    query: str
    collection: str
    total: int
    timings: QueryTimings | None = None


class MultiQueryRequest(BaseModel):
    query: str
    collections: list[str] = Field(min_length=1, max_length=20)
    top_k: int = Field(default=10, ge=1, le=1000)
    filters: dict | None = None
    min_score: float | None = None
    search_mode: str = Field(default="semantic", pattern=r"^(semantic|keyword|hybrid)$")
    rerank: bool | None = None


class MultiQueryResult(BaseModel):
    id: str
    text: str
    score: float
    document_id: str | None = None
    document_filename: str | None = None
    chunk_index: int | None = None
    collection: str = ""
    metadata: dict = {}


class MultiQueryResponse(BaseModel):
    results: list[MultiQueryResult]
    query: str
    collections: list[str]
    total: int


class BatchQueryItem(BaseModel):
    collection: str
    query: str
    top_k: int = Field(default=10, ge=1, le=1000)
    filters: dict | None = None
    min_score: float | None = None
    search_mode: str = Field(default="semantic", pattern=r"^(semantic|keyword|hybrid)$")
    rerank: bool | None = None


class BatchQueryRequest(BaseModel):
    queries: list[BatchQueryItem] = Field(min_length=1, max_length=20)


class BatchQueryResultItem(BaseModel):
    results: list[QueryResult]
    query: str
    collection: str
    total: int


class BatchQueryResponse(BaseModel):
    results: list[BatchQueryResultItem]


class EmbeddingModelInfo(BaseModel):
    provider: str
    model: str
    dimension: int
    description: str = ""


class EmbeddingModelListResponse(BaseModel):
    models: list[EmbeddingModelInfo]


class AnalyticsResponse(BaseModel):
    collection: str
    period_24h: dict
    period_7d: dict
    period_30d: dict
    top_queries: list[dict]

File: sdks/rust/src/types/query.rs
use serde::{Deserialize, Serialize};

/// Search mode for query operations.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SearchMode {
    /// Vector similarity search.
    Semantic,
    /// Keyword/BM25 search.
    Keyword,
    /// Combined semantic + keyword search.
    Hybrid,
}

/// Body for a single-collection query.
#[derive(Debug, Clone, Default, Serialize)]
pub struct QueryBody {
    /// Search query text (required).
    pub query: String,
    /// Maximum number of results.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_k: Option<u32>,
    /// Metadata filters.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filters: Option<serde_json::Value>,
    /// Minimum similarity score.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_score: Option<f64>,
    /// Search mode.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub search_mode: Option<SearchMode>,
    /// Whether to apply reranking.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rerank: Option<bool>,
}

/// A single query result.
#[derive(Debug, Clone, Deserialize)]
pub struct QueryResult {
    /// Result ID.
    pub id: String,
    /// Matched text content.
    pub text: String,
    /// Similarity score (0–1).
    pub score: f64,
    /// Source document ID.
    pub document_id: Option<String>,
    /// Source document filename.
    pub document_filename: Option<String>,
    /// Chunk index within the source document.
    pub chunk_index: Option<u32>,
    /// Source page number when available.
    pub page_no: Option<u32>,
    /// Source character start when available.
    pub char_start: Option<u32>,
    /// Source character end when available.
    pub char_end: Option<u32>,
    /// Result metadata.
    pub metadata: serde_json::Value,
}

/// Retrieval timings.
#[derive(Debug, Clone, Deserialize)]
pub struct QueryTimings {
    /// Embedding latency in milliseconds.
    pub embed_ms: f64,
    /// Search latency in milliseconds.
    pub search_ms: f64,
    /// Reranking latency in milliseconds.
    pub rerank_ms: f64,
    /// Cache lookup latency in milliseconds.
    pub cache_ms: f64,
    /// Total retrieval latency in milliseconds.
    pub total_ms: f64,
    /// Whether the retrieval results came from cache.
    pub cache_hit: bool,
}

/// Response from a single-collection query.
#[derive(Debug, Clone, Deserialize)]
pub struct QueryResponse {
    /// Matched results.
    pub results: Vec<QueryResult>,
    /// The query that was executed.
    pub query: String,
    /// Collection that was searched.
    pub collection: String,
    /// Total number of results.
    pub total: u32,
    /// Retrieval timings.
    pub timings: Option<QueryTimings>,
}

/// Body for a multi-collection query.
#[derive(Debug, Clone, Serialize)]
pub struct MultiQueryBody {
    /// Search query text (required).
    pub query: String,
    /// Collections to search across (required).
    pub collections: Vec<String>,
    /// Maximum number of results.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_k: Option<u32>,
    /// Metadata filters.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filters: Option<serde_json::Value>,
    /// Minimum similarity score.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_score: Option<f64>,
    /// Search mode.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub search_mode: Option<SearchMode>,
    /// Whether to apply reranking.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rerank: Option<bool>,
}

/// A single result from a multi-collection query, includes source collection.
#[derive(Debug, Clone, Deserialize)]
pub struct MultiQueryResult {
    /// Result ID.
    pub id: String,
    /// Matched text content.
    pub text: String,
    /// Similarity score.
    pub score: f64,
    /// Source document ID.
    pub document_id: Option<String>,
    /// Source document filename.
    pub document_filename: Option<String>,
    /// Chunk index within the source document.
    pub chunk_index: Option<u32>,
    /// Collection this result came from.
    pub collection: String,
    /// Result metadata.
    pub metadata: serde_json::Value,
}

/// Response from a multi-collection query.
#[derive(Debug, Clone, Deserialize)]
pub struct MultiQueryResponse {
    /// Matched results across collections.
    pub results: Vec<MultiQueryResult>,
    /// The query that was executed.
    pub query: String,
    /// Collections that were searched.
    pub collections: Vec<String>,
    /// Total number of results.
    pub total: u32,
}

/// A single query within a batch query request.
#[derive(Debug, Clone, Serialize)]
pub struct BatchQueryItem {
    /// Collection to query.
    pub collection: String,
    /// Search query text.
    pub query: String,
    /// Maximum number of results.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_k: Option<u32>,
    /// Metadata filters.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filters: Option<serde_json::Value>,
    /// Minimum similarity score.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_score: Option<f64>,
    /// Search mode.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub search_mode: Option<SearchMode>,
    /// Whether to apply reranking.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rerank: Option<bool>,
}

/// Body for a batch query request.
#[derive(Debug, Clone, Serialize)]
pub struct BatchQueryBody {
    /// Queries to execute (1–20).
    pub queries: Vec<BatchQueryItem>,
}

/// Response from a batch query.
#[derive(Debug, Clone, Deserialize)]
pub struct BatchQueryResponse {
    /// One `QueryResponse` per query in the batch.
    pub results: Vec<QueryResponse>,
}
